addons/phone_validation/lib/phonenumbers_patch/__init__.py

```diff
@@ -27,5 +27,17 @@ def _local_load_region(code):
         # loading updated region_CI.py from current directory
         # https://github.com/daviddrysdale/python-phonenumbers/blob/v8.12.32/python/phonenumbers/data/region_CI.py
         phonenumbers.phonemetadata.PhoneMetadata.register_region_loader('CI', _local_load_region)
+    # MONKEY PATCHING phonemetadata to fix Brazilian phonenumbers following 2016 changes
+    def _hook_load_region(code):
+        phonenumbers.data._load_region(code)
+        if code == 'BR':
+            phonenumbers.data.region_BR.PHONE_METADATA_BR.intl_number_format.append(
+                phonenumbers.phonemetadata.NumberFormat(
+                    pattern='(\\d{2})(\\d{4})(\\d{4})',
+                    format='\\1 9\\2-\\3',
+                    leading_digits_pattern=['(?:[14689][1-9]|2[12478]|3[1-578]|5[13-5]|7[13-579])'],
```

Comment:
The leading_digits_pattern is the pattern for mobile phone numbers (similar pattern can be found here https://github.com/daviddrysdale/python-phonenumbers/blob/v8.12.32/python/phonenumbers/data/region_BR.py )
Concerning the [2-57] part that we do not add: 
For the \d{2})(\d{4})(\d{4}) pattern there are 3 cases:
First case: the last 8 digits start with 2 to 5, it's a fixed phone number (we do not fix this one)
`fixed_line=PhoneNumberDesc(national_number_pattern='(?:[14689][1-9]|2[12478]|3[1-578]|5[13-5]|7[13-579])[2-5]\\d{7}', example_number='1123456789', possible_length=(10,), possible_length_local_only=(8,)),`
Second case: The last 8 digits starts with 7, it's an exception in the brazil phone system, we should not add a 9 to those numbers (iDEN mobile lines https://en.wikipedia.org/wiki/List_of_dialling_codes_in_Brazil )
Third case: it's the case that we add, we target mobile phone number (not leading with 2-5) and not mobile number leading with 7. Those numbers are invalid as of 2017 if they don't start by 9 before the last 8 digits, which we fix in this PR. The pattern (\\d{2})(\\d{4})(\\d{4}) ensures we fix the right phone numbers as well.

Reply:
as a side note, the valid case of 8 last digits begin with:

*  2-5 (landed line)
* or 7 (for some [iDEN radio thing](https://en.wikipedia.org/wiki/IDEN) "as of 2017 all mobile numbers in Brazil start with the digit 9. (There is an exception for some iDEN mobile lines operated by Nextel, which are eight digits long and start with 7 and disestablished in 2018.)"  from [wikipedia List_of_dialling_codes_in_Brazil](https://en.wikipedia.org/wiki/List_of_dialling_codes_in_Brazil))

are taken care by this rule:

```py
NumberFormat(pattern='(\\d{2})(\\d{4})(\\d{4})', format='\\1 \\2-\\3', leading_digits_pattern=['(?:[14689][1-9]|2[12478]|3[1-578]|5[13-5]|7[13-579])[2-57]']),
```

and the new added rule is exactly the same without `[2-57]` (so it's very similar to other entry from this list) so it will apply to phone numbers with the 8 last digits that begin with 1 or 6 or 8 or 9 (as the tests shows) since the first rules matches beginning by 2,3,4,5,7.

technically from wikipedia it seems mobiles were only beginning with  6 to 8 at the start of the 8 digits, so possibly we could add [68] at the end of the new rule if we wanted to be more restrictive

Reply:
Ok, it was a journey but I think I get how this thing works now thanks ^^'

From what I see it stops looking for matches the first time the pattern+leading_digits holds but indeed I'd say adding `6 through 9` to check it's a mobile number by standard shouldn't hurt.